### ISSUE TITLE
Consider a standardized way to modify hover/focus/active states for blocks

### ISSUE THREAD

**Issue by kjellr** [WordPress/gutenberg#27075]
Gutenberg should consider adding a way for users (and/or theme developers) to modify hover and active states. This would ideally be available in the UI on a per-block level, and also a part of global styles. This feature would be particularly useful for link styles and also for buttons. 

Related to https://github.com/WordPress/gutenberg/issues/4543

**Comment by karmatosed** [comment]
I would love to see more consistency in the way hover/focus/active states behave themselves, which to me leads to a more consistent way for them to be styled. I have been working on an example to refine the gallery block borders, which overlapped. Even a small change like that was a bit of a tangled thread to follow due to the nature of the block. Which makes me feel this is an opportunity to refine and standardize.

**Comment by MaggieCabrera** [comment]
I've been working on Seedlet's button hover states and I feel like the solution that I found shouldn't have to be replicated for every theme that is ever developed. Even if it's never surfaced to the user the ability to change the hover color/background should be handled by the editor to some capacity. In particular when the color is one that is selected by the user and not one provided by the theme.

For the particular case, I was considering that we could either play with the opacity of the background color (using rgba) or play with the lightness (using hsl). Could this be done directly within the editor? As it stands, a custom colored button doesn't have any hover state.

**Comment by carolinan** [comment]
Changing the color alone is not enough to indicate that the state of the link (button block) has changed, 
_unless_ the color contrast ratio between the two colors is significant. What the minimum contrast ratio is also depends on factors like the font size.  https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/#color-contrast

If there is a hover color option exposed to users, it may need to be a selection of suggested accessible colors, based on the current button color(s), and refreshed if the user changes the button color.

**Comment by Tropicalista** [comment]
I added recently an issue: https://github.com/WordPress/gutenberg/issues/30068

I suppose that `hover` is not integrated yet because there's no way to add pseudo class on inline style. A different approach to add custom hover effect should be proposed.

**Comment by getdave** [comment]
I'm going to add this for discussion in the Core Editor chat. It's something we are going to [need for the Navigation block](https://github.com/WordPress/gutenberg/issues/38277).

**Comment by aurooba** [comment]
So I think one of the considerations is _what_ changes with each `state`. A `state` change isn't just a change of `color` or `text-decoration`, in CSS speak. It could be multiple things. 

I would suggest that for each state, all currently supported properties remain available, so if the block supports `background color`, `margin`, `padding`, and `letter case`, then you should be able to control each of those per `state`. This will likely need some CSS adjustments, but should be pretty achievable with CSS variables.

**Comment by joyously** [comment]
> I would suggest that for each state, all currently supported properties remain available, so if the block supports `background color`, `margin`, `padding`, and `letter case`, then you should be able to control each of those per `state`.

I disagree because of what can happen when hover changes the size of an element. It has been seen before in the editor that just hovering creates a scrollbar or other size change, which then moves the element so it's no longer hovered, so it changes size again. This creates a flickering of the element that can cause seizures. Even a little movement is annoying.

I don't think the editor should support the hover, active, focus, visited except on links. If it's standardized, the correct order is maintained. Be aware that visited can only affect color, though. I'm in favor of leaving it to the theme, not the user.

**Comment by aurooba** [comment]
> It has been seen before in the editor that just hovering creates a scrollbar or other size change, which then moves the element so it's no longer hovered, so it changes size again. This creates a flickering of the element that can cause seizures. Even a little movement is annoying.

I think there would be value in considering _when_ states show up. Something I love about Adobe XD for example is that, yes you can set a hover state for an element, but you have to toggle that state _on_ in the interface to see what it will look like, instead of having it appear whenever you _physically_ hover on the element. That's a better experience than what we have now, where if something has a hover state it activates in the editor on hover. Something like this would address the concern you are pointing out @joyously, which is absolutely valid.

> I don't think the editor should support the hover, active, focus, visited except on links.

I don't think it makes sense to support _states_ on everything. But I do think more than just links should support it. The blocks (off the top of my head) that I would love to see _state_ support for are (aside from just links anywhere): buttons, social links, file, site title, and navigation.

**Comment by joyously** [comment]
> That's a better experience than what we have now, where if something has a hover state it activates in the editor on hover. Something like this would address the concern you are pointing out

That's not really what I meant. I was only referring to the front end result of whatever the user chose to change on hover. Their visitors could be faced with lots of issues that are avoided if controls are not provided to the user. That's why I think themes should handle it instead.
> The blocks (off the top of my head) that I would love to see state support for are (aside from just links anywhere): buttons, social links, file, navigation, site title, and navigation.

Most of those are links (even navigation mentioned twice).

**Comment by aurooba** [comment]
Right, but themes can do all kinds of things on the front-end as well (and I've seen plenty that to), ultimately what kind of styles/changes are made on hover or any other state is up to the people curating the design/look. I don't think that should stop the editor from providing flexible tools.

And yes, most of the blocks I mentioned are links, but I'd like to see states available in them specifically, and not just link control on a global level. :)


Edit: the reason I am advocating for this approach is because I believe this stuff should be in theme.json. And, correct me if I'm wrong, but if it's in the editor, then it'll also be in theme.json.

**Comment by andrewserong** [comment]
Great discussion, folks!

> The blocks (off the top of my head) that I would love to see state support for are (aside from just links anywhere): buttons, social links, file, site title, and navigation.

I think this is a good argument for coming up with a generic way that multiple states can be handled, both at a block level and within `theme.json` so that Gutenberg can support these kinds of states for blocks. We'd then be able to look into where and how the controls would be exposed (i.e. some themes might not want users to be able to adjust directly within the post editor, while others might want to give more freedom, but for designing a theme the site editor at least should be able to support it?).

It sounds like there's possibly two different parts that need to be explored for this problem: 1) the design of how a user modifies the state in global styles or at an individual block level, and 2) how we store that styling state and render it out. I've added a couple of notes about the second point in the more recent related issue: https://github.com/WordPress/gutenberg/issues/38277#issuecomment-1028498013

**Comment by anarieldesign** [comment]
> Gutenberg should consider adding a way for users (and/or theme developers) to modify hover and active states. This would ideally be available in the UI on a per-block level, and also a part of global styles. This feature would be particularly useful for link styles and also for buttons.
> 
> Related to #4543

I totally agree with, if user can change color, border etc. for the button it should be available to change the hover, active states. For example when creating different button block patterns in a theme u need to style a generic hover effect for each button or to use different Custom CSS for each one to style it differently and when user change the style of that button the hover will still be the one from the theme. In short if you offer a color and styles for the button I think we need to have styles for the hover, active state

**Comment by MaggieCabrera** [comment]
> I totally agree with, if user can change color, border etc. for the button it should be available to change the hover, active states. For example when creating different button block patterns in a theme u need to style a generic hover effect for each button or to use different Custom CSS for each one to style it differently and when user change the style of that button the hover will still be the one from the theme. In short if you offer a color and styles for the button I think we need to have styles for the hover, active state

+1, we can discuss about UI and what options the end user will see in the editor, but the theme developer should be able to control and set default to those attributes on all the states without having to add extra CSS

**Comment by aurooba** [comment]
@andrewserong Does that mean this issue should be closed and all discussion should move to #38277?

**Comment by getdave** [comment]
> @andrewserong Does that mean this issue should be closed and all discussion should move to #38277?

Just noting that #38277 is _specifically_ related to the Nav block as it's part of the Nav block tracking Issue.

**Comment by andrewserong** [comment]
> Does that mean this issue should be closed and all discussion should move to...

I think it's fine to have both issues open. This one seems like a good place for the more general discussion, and it's also the earlier issue 😀

**Comment by Luehrsen** [comment]
Just to chime in with a real world example, this is how we approach states in our client themes at the moment.

```css
.has-primary-color {
	--current-color: var(--color-primary);
	--current-hover-color: var(--color-primary-hover);
	--current-contrast-color: var(--color-white);
	color: var(--color-primary);
}

.has-primary-background-color {
	--current-background-color: var(--color-primary);
	--current-hover-background-color: var(--color-primary-hover);
	--current-contrast-background-color: var(--color-white);
	background-color: var(--color-primary);

	&.has-background-dim {
		background-color: var(--color-primary);
	}
}
```

These variables are then picked up and combined in the elements that have state and are used accordingly.

**Comment by scruffian** [comment]
Right, but this should really be handled by Gutenberg rather than every theme having to maintain their own version of this code.

**Comment by joyously** [comment]
> but this should really be handled by Gutenberg rather than every theme having to maintain their own version of this code.

What is a theme for, then? 
I don't think the user should be involved, so there should be no UI for it, so the editor should not care what the theme does with those styles.

**Comment by aurooba** [comment]
> I don't think the user should be involved, so there should be no UI for it, so the editor should not care what the theme does with those styles.

I disagree. The option for the user to be involved should be there, _with_ the ability for theme authors to disable user customization of this feature if they wish. 

That is how the Block Editor has operated since being merged into core, and long term, that is the best approach. Provide the most flexibility and add curation/locking/limiting options that theme authors can utilize (I want to emphasize that flexibility without the ability to curate/lock/limit is terrible, and features should be developed in tandem with their ability to be locked down). This is important because a very large part of the WordPress end user community are DIY-ers who often start with a theme and then customize on top of it. I'm not a DIYer, but even if I find this helpful when I'm working on personal projects and need to do modifications quickly.

**Comment by Luehrsen** [comment]
> Right, but this should really be handled by Gutenberg rather than every theme having to maintain their own version of this code.

This implies that Gutenberg could handle every style a designer throws at a frontend developer. Considering Gutenberg competitors with a far superior scope (like Webflow, Squarespace, Wix), this is simply not the case. You will always be locked in by the capabilities of what the editor team deems worthy and would stifle innovation and creativity outside of what the team anticipated.

**Comment by paaljoachim** [comment]
@critterverse Channing shared a link to @javierarce post here which is interesting to share into this issue: http://work.javierarce.com/2021/09/14/interaction-of-color/

**Comment by getdave** [comment]
Thank you for the varied opinions expressed here. It's really great to see this discussion evolving.

For my part, I'd like to add that I don't see this as a binary choice. 

My opinion is that handling some _basic_ interaction styles (such as color) would be appropriate for the editor to handle. Indeed, many users will likely want to be able to have some control over these most basic of design choices.

That doesn't mean however, that Themes won't still be able to augment or override those settings via CSS to provide additional creativity. We could even have an option within `theme.json` to disable the ability to edit these interaction states.

What we don't want is anyone to be "locked in" or for any creativity to be stifled. I hope what I'm suggesting above addresses at least some of those concerns.

**Comment by oandregal** [comment]
Braindumping some thoughts regarding a particular detail we should address is how user choices affect states.

A theme may want to set the blue color for the rest state and the yellow one for hover. Is the user sets the rest color to a yellowish base, is it still good to have yellow for hover? A related though is that different states may require any CSS property to update, like for example padding or margin, as [in this use case](https://github.com/WordPress/gutenberg/issues/34141#issuecomment-952895572). In this sense, states are more like "block style variations" than they are "style properties".

One of the issues is that the theme lacks information about whether an user has styled anything. Or rather, there's no way for a theme to "maintain the design links" when the user edits something. I've shared a [potential direction to explore](https://github.com/WordPress/gutenberg/pull/39432#issuecomment-1067964350).

**Comment by cbirdsong** [comment]
> A related though is that different states may require any CSS property to update, like for example padding or margin, as [in this use case](https://github.com/WordPress/gutenberg/issues/34141#issuecomment-952895572). In this sense, states are more like "block style variations" than they are "style properties".

Updating properties that affect layout on hover/focus is [generally considered a very bad idea](https://web.dev/animations-overview/#layout). Would we really want to encourage that?

**Comment by getdave** [comment]
For anyone watching this Issue you might like to keep an eye on https://github.com/WordPress/gutenberg/issues/38277 which is where design progress is being made on this problem.

**Comment by talldan** [comment]
I'm not sure if this is still in progress, but the PR was closed so I've removed the assignment and In Progress label.

> For anyone watching this Issue you might like to keep an eye on https://github.com/WordPress/gutenberg/issues/38277 which is where design progress is being made on this problem.

@getdave From a triage point of view, what do you think about rewording the title and description of #38277? It seems like the conversation is no longer about the nav block, and it has become a duplicate of this issue. Potentially this issue could be closed or become an overview while #38277 becomes the design exploration.

I also don't think we want separate issues for adding hover states to individual blocks because that would become unmanageable. I just closed https://github.com/WordPress/gutenberg/issues/42829, but I think it's confusing for the reporter that I'm pointing them to an issue for the nav block.

**Comment by getdave** [comment]
I agree @talldan. Let's close this one in favour of https://github.com/WordPress/gutenberg/issues/38277 which, for whatever reason, has become canonical.

> I'm not sure if this is still in progress, but the PR was closed so I've removed the assignment and In Progress label.

It's still on my radar but I've switched focus to the Nav block for a while.
